### DISCUSSION TITLE
egui plot how to increase GPU usage instead of CPU

### DISCUSSION THREAD

**Discussion by GunnarMorrigan** [emilk/egui#3810] [Q&A]
Hi,

I would like to render my application on the GPU as this is much faster then CPU. 
My laptop has both intel iGPU Xe and a nvidia T600. The drivers are up to date. DirectX 12, with latest intel and nvidia drivers.

When I run my application which is just the demo application stripped with an egui_plot plot with 500k points the CPU spikes to 50% while the GPUs seems untouched. Frame rates get very choppy.

I would like to take more advantage of my GPU. 

Two questions:
- How can I know when the GPU is not used?
- How can I increase GPU usage?

**Comment by YgorSouza** [comment]
If you are plotting 500k points, the bottleneck is probably the allocation of the plot points, which has to be done by the CPU. See discussion in #1485.

**Comment by GunnarMorrigan** [comment]
Hi @YgorSouza,

Thanks for the reply. I adjusted the plot to take a mutable reference of the points but it still hits the CPU really hard.

The code is not polished but it works and takes a reference to the data only creating it once
https://github.com/GunnarMorrigan/egui/blob/5c43d23be6a5abd0cab5419c2874e8d47d2afcc0/crates/egui_plot/src/lib.rs#L712

I am not sure what is hitting the CPU so hard.
I also have never profiled Rust code before. Is there an easy and simple profiler to use?

**Comment by GunnarMorrigan** [comment]
I looked deeper into the problem. 
There is a large amount of time going to the reallocation of vectors.
There are two places that take the most time:
- Tessellation, with such a large number of shapes to draw, tessellation is a huge chunk of the work that needs to be done. Most of the problem is the vector in mesh not having sufficient capacity and requiring growth. 
I attempted to mitigate this by performing a reserve before the pushes here https://github.com/GunnarMorrigan/egui/blob/784f1b7e25228c5b9fd30047af32c331c21b3f4d/crates/epaint/src/mesh.rs#L198 this increases performance slightly but as 500k points need to be added to the mesh it will call reserve **a lot**.

- Turning points into shapes in the plot here: https://github.com/GunnarMorrigan/egui/blob/784f1b7e25228c5b9fd30047af32c331c21b3f4d/crates/egui_plot/src/items/mod.rs#L914C1-L921C22
This also does a lot of allocations with many points. It would be better to reserve a number of points beforehand. The number of pushes is known so I will adjust my fork to perform this check. 

@emilk do you possibly have an idea on how to optimize this further?
I have done profiling already but would like to know if you think there is any low hanging fruit.
